RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducers.cc

```diff
@@ -0,0 +1,164 @@
+    const device::EDPutToken<hgcalrechit::HGCalRecHitDevice> recHitsToken_;
+    HGCalRecHitCalibrationAlgorithms calibrator_;  // cannot be "const" because the calibrate() method is not const
+    int n_hits_scale;
+  };
+
+  HGCalRecHitsProducer::HGCalRecHitsProducer(const edm::ParameterSet& iConfig)
+      : digisToken_{consumes<hgcaldigi::HGCalDigiHost>(iConfig.getParameter<edm::InputTag>("digis"))},
+        recHitsToken_{produces()},
+        calibrator_{HGCalRecHitCalibrationAlgorithms(iConfig.getParameter<int>("n_blocks"),
+                                                     iConfig.getParameter<int>("n_threads"))},
+        n_hits_scale{iConfig.getParameter<int>("n_hits_scale")} {
+    calibToken_ = esConsumes(iConfig.getParameter<edm::ESInputTag>("calibSource"));
+    configToken_ = esConsumes(iConfig.getParameter<edm::ESInputTag>("configSource"));
+  }
+
+  void HGCalRecHitsProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("digis", edm::InputTag("hgcalDigis", "DIGI", "TEST"));
+    desc.add("calibSource", edm::ESInputTag{})->setComment("Label for calibration parameters");
+    desc.add("configSource", edm::ESInputTag{})->setComment("Label for ROC configuration parameters");
+    desc.add<int>("n_blocks", -1);
+    desc.add<int>("n_threads", -1);
+    desc.add<int>("n_hits_scale", -1);
+    descriptions.addWithDefaultLabel(desc);
+  }
+
+  void HGCalRecHitsProducer::beginRun(edm::Run const& iRun, edm::EventSetup const& iSetup) {
+    // auto moduleInfo = iSetup.getData(moduleInfoToken_);
+    // std::tuple<uint16_t,uint8_t,uint8_t,uint8_t> denseIdxMax = moduleInfo.getMaxValuesForDenseIndex();
+    // calibrationParameterProvider_.initialize(HGCalCalibrationParameterProviderConfig{.EventSLinkMax=std::get<0>(denseIdxMax),
+    //                                   .sLinkCaptureBlockMax=std::get<1>(denseIdxMax),
+    //                                   .captureBlockECONDMax=std::get<2>(denseIdxMax),
+    //                                   .econdERXMax=std::get<3>(denseIdxMax),
+    //                                   .erxChannelMax = 37+2,//+2 for the two common modes
+    // });
+  }
+
+  void HGCalRecHitsProducer::produce(device::Event& iEvent, device::EventSetup const& iSetup) {
+    auto queue = iEvent.queue();
+
+    // Read digis
+    auto const& deviceCalibParamProvider = iSetup.getData(calibToken_);
+    auto const& deviceConfigParamProvider = iSetup.getData(configToken_);
+    auto const& hostDigisIn = iEvent.get(digisToken_);
+
+// Check if there are new conditions and read them
+#ifdef EDM_ML_DEBUG
+    if (calibWatcher_.check(iSetup)) {
+      for (int i = 0; i < deviceConfigParamProvider.view().metadata().size(); i++) {
+        LogDebug("HGCalCalibrationParameter") << "gain = " << deviceConfigParamProvider.view()[i].gain();
+      }
+    }
+#endif
```

Comment:
This could be
```suggestion
    LogDebug("HGCalCalibrationParameter").log([&](auto& log) {
      if (calibWatcher_.check(iSetup)) {
        for (int i = 0; i < deviceConfigParamProvider.view().metadata().size(); i++) {
          log << "gain = " << deviceConfigParamProvider.view()[i].gain() << "\n";
        }
      }
    });
```
Same for the other printout below.

Reply:
+1